pymc3/smc/sample_smc.py

```diff
@@ -0,0 +1,161 @@
+    -----
+    SMC works by moving through successive stages. At each stage the inverse temperature
+    :math:`\beta` is increased a little bit (starting from 0 up to 1). When :math:`\beta` = 0
+    we have the prior distribution and when :math:`\beta` =1 we have the posterior distribution.
+    So in more general terms we are always computing samples from a tempered posterior that we can
+    write as:
+
+    .. math::
+
+        p(\theta \mid y)_{\beta} = p(y \mid \theta)^{\beta} p(\theta)
+
+    A summary of the algorithm is:
+
+     1. Initialize :math:`\beta` at zero and stage at zero.
+     2. Generate N samples :math:`S_{\beta}` from the prior (because when :math `\beta = 0` the
+         tempered posterior is the prior).
+     3. Increase :math:`\beta` in order to make the effective sample size equals some predefined
+        value (we use :math:`Nt`, where :math:`t` is 0.5 by default).
+     4. Compute a set of N importance weights W. The weights are computed as the ratio of the
+        likelihoods of a sample at stage i+1 and stage i.
+     5. Obtain :math:`S_{w}` by re-sampling according to W.
+     6. Use W to compute the covariance for the proposal distribution.
+     7. For stages other than 0 use the acceptance rate from the previous stage to estimate the
+        scaling of the proposal distribution and `n_steps`.
+     8. Run N Metropolis chains (each one of length `n_steps`), starting each one from a different
+        sample in :math:`S_{w}`.
+     9. Repeat from step 3 until :math:`\beta \ge 1`.
+     10. The final result is a collection of N samples from the posterior.
+
+
+    References
+    ----------
+    .. [Minson2013] Minson, S. E. and Simons, M. and Beck, J. L., (2013),
+        Bayesian inversion for finite fault earthquake source models I- Theory and algorithm.
+        Geophysical Journal International, 2013, 194(3), pp.1701-1726,
+        `link <https://gji.oxfordjournals.org/content/194/3/1701.full>`__
+
+    .. [Ching2007] Ching, J. and Chen, Y. (2007).
+        Transitional Markov Chain Monte Carlo Method for Bayesian Model Updating, Model Class
+        Selection, and Model Averaging. J. Eng. Mech., 10.1061/(ASCE)0733-9399(2007)133:7(816),
+        816-832. `link <http://ascelibrary.org/doi/abs/10.1061/%28ASCE%290733-9399
+        %282007%29133:7%28816%29>`__
+    """
+
+    smc = SMC(
+        draws=draws,
+        kernel=kernel,
+        n_steps=n_steps,
+        parallel=parallel,
+        start=start,
+        cores=cores,
+        tune_steps=tune_steps,
+        p_acc_rate=p_acc_rate,
+        threshold=threshold,
+        epsilon=epsilon,
+        dist_func=dist_func,
+        sum_stat=sum_stat,
+        progressbar=progressbar,
+        model=model,
+        random_seed=random_seed,
+    )
+
+    _log = logging.getLogger("pymc3")
+    _log.info("Sample initial stage: ...")
+    stage = 0
+    smc.initialize_population()
+    smc.setup_kernel()
+    smc.initialize_logp()
+
+    while smc.beta < 1:
```

Comment:
Risk of it running forever? Maybe we should have an additional termination rule like after N iterations.

Reply:
Early stopping will return a non useful result as the samples will not be from the posterior. Thus I think is OK to let the user kill the process if it is running more than what the user can wait. One alternative will be to jump to beta=1 after N iterations, but most likely than not that will also give a poor sample